Make directory deletion more atomic

mikepavone opened the pull request:
Currently, if a file is locked in the destination directory (which will be the case if the app is running), it will end up partially deleted and the installation will exit with an error. This leaves the existing install in a corrupted state, which is less than ideal. This PR makes deletes sort-of atomic by first renaming the directory before deleting it. This rename will fail if the app is currently running or if the contents of the directory happen to be locked for some other reason

mikepavone committed:
Make directory deletion more atomic

dyc commented on src/Squirrel/Utility.cs:
```diff
@@ -388,6 +406,12 @@ public static async Task DeleteDirectory(string directoryPath)
                 Log().ErrorException(message, ex);
+                if (moveFirst) {
+                    // move back to original name so we can try again on the next run
+                    Directory.Move(directoryPath, originalPath);
```
interesting. so `Move()` throws and then we're left with a partially moved directory that this second call restores, presumably only touching files that weren't locked & unable to be moved the first time?


and then how does this manifest to the user when they run `Setup.exe`, as a dialog as i think was mentioned somewhere?

mikepavone replied:
when DeleteDirectory throws during the install action, you get a generic error modal with a button that opens the setup log file that contains the actual exception

if the initial `Move()` throws, nothing will have been modified (which is the goal here). if this `Move()` fails we have a stickier situation as it means we have failed to fully delete the directory and we weren't able to rename it back to the original name. So we will effectively be leaving some garbage behind in the user's appdata directory in this case. The hope is that this will be a rare failure in practice since the whole point of doing the initial `Move()` is that it will fail if any of the children are locked and so the recursive delete will hopefully succeed in full. Even if the delete fails, the hope is that if we were able to rename the directory once, we'll be able to do it again.

adill reviewed (commented):
applying this to every `DeleteDirectory` seems like wielding a very large hammer. i don't have full context here but i feel like we would do ourselves a favor by specifically targeting whichever problematic case we're attempting to address.

it looks like *attempts were made* to handle exceptions inside this function and manifest them as error logs. we're essentially adding a new method of failure (a `Directory.Move` exception) that i'm not convinced would be well-received by all the callers.

wdyt about wrapping the first `Directory.Move` in a try/catch and just logging & returning in the catch? similarly, when attempting to restore state after a failed delete... what should happen if the move-back fails?

mikepavone committed:
Some changes in response to review feedback

mikepavone commented:
> applying this to every `DeleteDirectory` seems like wielding a very large hammer. i don't have full context here but i feel like we would do ourselves a favor by specifically targeting whichever problematic case we're attempting to address.

Fair enough. I've changed the default value of `moveFirst` and only explicitly set it to true when attempting to delete the top-level destination directory in the full install action. The original thinking was that "atomic" deletes are a much more sensible default behavior, but looking at the other use cases partial deletes seem desirable in at least some of the usages (removing temporary directories created for unpacking artifacts for install).

> it looks like attempts were made to handle exceptions inside this function and manifest them as error logs. we're essentially adding a new method of failure (a Directory.Move exception) that i'm not convinced would be well-received by all the callers.
> 
> wdyt about wrapping the first Directory.Move in a try/catch and just logging & returning in the catch?

In practice, this already throws in the most common error case (or at least what I assume to be our most common error case) because File.DeleteFile is not protected by a try/catch. It's not entirely clear to me why not throwing is desirable when the directory deletion is failed. Unfortunately that seems to be a decision that predates the current upstream repo.

Regardless, the throw on the initial move failure is needed for the "target" case. We need it to throw so it aborts the installation with an error rather than potentially corrupting the current install with a half applied new one. With the change above, it will only potentially throw in this case when we explicitly want it (at least now that I removed the `throw ex;` I added to the last `catch` block)

> similarly, when attempting to restore state after a failed delete... what should happen if the move-back fails?

There's not really a good way to make sure the folder ultimately gets cleaned up in this case. I guess we could store the temp folder name somewhere and try to clean it up later, but that feels pretty gross. 

Fortunately, I think this situation should be rare. I guess thinking about it more, it doesn't make sense to throw in this case as the target directory has been moved out of the way even if it's not completely gone

adill reviewed (approved):
👍🏻
